subprojects/core/src/main/java/org/gradle/configuration/internal/DefaultDynamicCallContextTracker.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.configuration.internal;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+import java.util.Stack;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.Consumer;
+
+public class DefaultDynamicCallContextTracker implements DynamicCallContextTracker {
+    private static class State {
+        Stack<Object> entryPointsStack = new Stack<>();
+    }
+
+    private final ThreadLocal<State> stateByThread = ThreadLocal.withInitial(State::new);
+    private final List<Consumer<Object>> enterListeners = new CopyOnWriteArrayList<>();
```

Comment:
There should normally be just one of each kind, and they are added during the service registration. That's why I decided to avoid the synchronization overhead by using `CopyOnWriteArrayList`.